apis/v1beta1/config.go

```diff
@@ -441,67 +441,119 @@ const (
+		if len(explicitPortMatches) <= 1 {
+			return telemetry.Metrics.Address, defaultServicePort, nil
+		}
 
-	port, err := strconv.ParseInt(explicitPortMatches[1], 10, 32)
-	if err != nil {
-		errMsg := fmt.Sprintf("couldn't determine metrics port from configuration: %s",
-			telemetry.Metrics.Address)
-		logger.Info(errMsg, "error", err)
-		return "", 0, err
-	}
+		port, err := strconv.ParseInt(explicitPortMatches[1], 10, 32)
+		if err != nil {
+			errMsg := fmt.Sprintf("couldn't determine metrics port from configuration: %s",
+				telemetry.Metrics.Address)
+			logger.Info(errMsg, "error", err)
+			return "", 0, err
+		}
 
-	host, _, _ := strings.Cut(telemetry.Metrics.Address, explicitPortMatches[0])
-	return host, int32(port), nil
+		host, _, _ := strings.Cut(telemetry.Metrics.Address, explicitPortMatches[0])
+		return host, int32(port), nil
+	}
+
+	port := defaultServicePort
+	host := defaultServiceHost
+	for _, reader := range telemetry.Metrics.Readers {
+		if reader.Pull != nil {
+			if reader.Pull.Exporter != nil {
+				if reader.Pull.Exporter.Prometheus != nil {
+					if reader.Pull.Exporter.Prometheus.Host != "" {
+						host = reader.Pull.Exporter.Prometheus.Host
+					}
+					if reader.Pull.Exporter.Prometheus.Port != 0 {
+						port = reader.Pull.Exporter.Prometheus.Port
+					}
+					break
+				}
+			}
+		}
+	}
+	return host, port, nil
 }
 
 // ApplyDefaults inserts configuration defaults if it has not been set.
 func (s *Service) ApplyDefaults(logger logr.Logger) error {
-	telemetryAddr, telemetryPort, err := s.MetricsEndpoint(logger)
-	if err != nil {
-		return err
-	}
 
-	tm := &AnyConfig{
+	reader := s.getPrometheusMetricReader(logger)
+	tm := AnyConfig{
 		Object: map[string]interface{}{
 			"metrics": map[string]interface{}{
-				"address": fmt.Sprintf("%s:%d", telemetryAddr, telemetryPort),
+				"readers": []any{reader},
 			},
 		},
 	}
 
 	if s.Telemetry == nil {
-		s.Telemetry = tm
+		s.Telemetry = &tm
 		return nil
 	}
-	// NOTE: Merge without overwrite. If a telemetry endpoint is specified, the defaulting
-	// respects the configuration and returns an equal value.
+
+	// This field is deprecated and should not be used anymore.
+	tel := s.GetTelemetry()
+	if tel.Metrics.Address != "" {
+		// Setting the "address" field to an empty string explicitly removes the value
+		// during Kubernetes serialization. Directly deleting the field from the map using
+		// delete(metrics, "address") does not work because Kubernetes treats missing fields
+		// differently from explicitly empty ones. By assigning "", we ensure the configuration
+		// is updated correctly when the resource is persisted.
+		s.Telemetry.Object["metrics"].(map[string]interface{})["address"] = ""
+	}
+
 	if err := mergo.Merge(s.Telemetry, tm); err != nil {
 		return fmt.Errorf("telemetry config merge failed: %w", err)
 	}
+
 	return nil
 }
 
+func (s *Service) getPrometheusMetricReader(logger logr.Logger) any {
+	getReader := func(addr string, port int32) any {
```

Comment:
why does this need to be a function?

Reply:
Before I was doing something more complicated and made more sense... I will simplify.